block-layer/content-addressable-archives.md

```diff
@@ -30,51 +30,110 @@
+
+### Version 1
+
+A version 1 CAR combines the version block and header block into a single DAG-CBOR Header block at the begining of the format:
 
 ```
 |--------- Header --------| |---------------------------------- Data -----------------------------------|
 
 [ varint | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] [ varint | CID | block ] …
 ```
 
-### Header
+As with Version 2, the initial varint is read then the first block is read from the number of bytes determined by the initial varint. This block is the CAR _Header_, a DAG-CBOR (CBOR with tag `42` for CIDs) encoded object holding the version number and array of roots.
 
-The first bytes of the CAR format hold a varint, this unsigned integer specifies the number of bytes beyond the varint itself that contain the _Header_ block. This Header block is a byte array DAG-CBOR (CBOR with tag 42 for CIDs) encoded object holding the version number and array of roots. As an [IPLD Schema](../schemas/):
+Described as an [IPLD Schema](../schemas/):
 
 ```ipldsch
-type CarHeader struct {
+type CarV1Header struct {
   version Int
   roots [&Any]
 }
 ```
 
+All trailing blocks constitute the Data section of the CAR, where blocks are prefixed by a varint describing the number of bytes to read, followed by the block's CID, followed by the block's binary data.
+
 #### Constraints
 
-* The `version` is always a value of `1`. Future iterations of this specification may make use of `version` to introduce variations of the format.
-* The `roots` array must contain **one or more** CIDs, each of which should be present somewhere in the remainder of the CAR.
+* The `roots` array should contain **one or more** CIDs, each of which should be present somewhere in the remainder of the CAR. A zero-length `roots` array is possible but not recommended due to assumptions present in downstream code.
+
+### Version 2
 
-_(Caveats: see [Number of roots](#number-of-roots) and [Root CID block existence](#root-cid-block-existence) under Unresolved Issues.)_
+```
+|------ Version ------| |----------- Header ------------| |---------------------- Data ----------------------|
+
+[ varint | cbor block ] [ varint | CID | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] …
+```
+
+To maintain backward compatibility, Version 2 retains the lead-in block without CID but _only_ includes a CBOR encoded map containing a single key/value pair: `{ version: 2 }`. Therefore, a decoder may read the first 11 bytes of a CAR and match it against the bytes: `0x0aa16776657273696f6e02` (`0x0a` is a varint representing `10`, followed by 10 bytes of the CBOR encoded form of `{ version: 2 }`).
```

Comment:
I think this was actually my idea.

One way we could maintain backwards compatibility is to keep the old v1 style header but only ever encode it as `{ version: 2 }`. To the old v1 parser this looks like a CAR file with no roots. For a v2 parser you don’t need to even parse it, you can check if it’s a v2 CAR file by doing a simple binary comparison.

Reply:
Step a little further back though: if we accept that v1 files "in the wild" all start with `0x0aa16776657273696f6e02`, then the only concern you have is "what will a v1 ( old) parser do when it encounters a `ver != 1` file. We want it to croak. Having v2 being a uniform stream of `{cidlen}{cid}{payloadlen}{payload}` should not possibly be decodeable as valid cbor, hence "problem solved"? Or you are concerned about the possibility of crafting a valid v2 file that "means something" to a v1 parser...?

Reply:
> We want it to croak

There are many different forms of “croak” though 😉

Being unable to even read the header is likely to cause an exception. Reading the header, seeing that the version is beyond the available range, and not having any root information to use in order to parse the file, is a much less aggressive form of croaking. It’s likely to cause the CAR file to simply be ignored rather than causing a noticeable error.

Now, this may not be the desired behavior. We may *want* a stronger form of croaking, we may want to see clear exceptions, but I’m not sure we do and I’m a little worried about code in the Filecoin ecosystem encountering new CAR files in the future and what the failure mode looks like.

Reply:
> Being unable to even read the header is likely to cause an exception.

Only in clients linked against a sufficiently old go-car not knowing how to read v != 1

Or are you saying that at this point there is likely a significant number of implementations out there that wrote their own car handlers not using any of our libraries?

Reply:
I have a high degree of certainty that, by the time we ship this, there will be car handlers we didn’t write in use. For one thing, we don’t have a Rust CAR implementation and there’s Rust stuff happening in the ecosystem.

How well they are written and if they can handle something like this, I have no idea.

Reply:
I see...
So perhaps we should pivot and ship v2 with **only** the following changes:
- make versioning forwards compatible
- potentially fix the cid-length specification / symmetry of data-frames questions

And leave everything else for v3?

Reply:
I'm not sure we're willing to drop selectors from scope, are we? We already have Go code in go-car that's passing in a Root+Selector (https://github.com/ipld/go-car/blob/master/selectivecar.go#L21-L24) and presumably corresponding code in go-fil-markets that's working this way, so the data is there and we could get ontop of this pretty quick.